crates/holochain_types/src/db_cache.rs

```diff
@@ -0,0 +1,575 @@
+            .get_or_try_init(|| {
+                let env = self.dht_env.clone();
+                async move {
+                    let (activity_integrated, mut all_activity) = env
+                        .async_reader(|txn| {
+                            // Get the highest integrated sequence number for each agent.
+                            let activity_integrated: Vec<(AgentPubKey, u32)> = txn
+                            .prepare_cached(
+                                holochain_sqlite::sql::sql_cell::ACTIVITY_INTEGRATED_UPPER_BOUND,
+                            )?
+                            .query_map(
+                                named_params! {
+                                    ":register_activity": DhtOpType::RegisterAgentActivity,
+                                },
+                                |row| {
+                                    Ok((
+                                        row.get::<_, Option<AgentPubKey>>(0)?,
+                                        row.get::<_, Option<u32>>(1)?,
+                                    ))
+                                },
+                            )?
+                            .filter_map(|r| match r {
+                                Ok((a, seq)) => Some(Ok((a?, seq?))),
+                                Err(e) => Some(Err(e)),
+                            })
+                            .collect::<rusqlite::Result<Vec<_>>>()?;
+
+                            // Get all the agents that have activity ready to be integrated.
+                            let all_activity_agents: Vec<Arc<AgentPubKey>> = txn
+                                .prepare_cached(
+                                    holochain_sqlite::sql::sql_cell::ALL_ACTIVITY_AUTHORS,
+                                )?
+                                .query_map(
+                                    named_params! {
+                                        ":register_activity": DhtOpType::RegisterAgentActivity,
+                                    },
+                                    |row| Ok(Arc::new(row.get::<_, AgentPubKey>(0)?)),
+                                )?
+                                .collect::<rusqlite::Result<Vec<_>>>()?;
+
+                            // Any agent activity that is currently ready to be integrated.
+                            let mut any_ready_activity: HashMap<Arc<AgentPubKey>, ActivityState> =
+                                HashMap::with_capacity(all_activity_agents.len());
+                            let mut stmt = txn.prepare_cached(
+                                holochain_sqlite::sql::sql_cell::ALL_READY_ACTIVITY,
+                            )?;
+
+                            // For each agent with activity that is ready to be integrated gather all
+                            // the chain items and add them to the `out_of_order` list.
+                            for author in all_activity_agents {
+                                let out_of_order = stmt
```

Comment:
It's just a `Vec<u32>` so unless someone was intentionally holding back an upstream chain item then this can't happen but chain items require the previous header to validate so these would be held up in validation anyway.